### DISCUSSION TITLE
Using Rust to not have to touch Yaml in k8s land

### DISCUSSION THREAD

**Discussion by dpc** [rustshop/rustshop#12] [Technical discussion]
![Yaml is lava!](https://i.imgflip.com/6mefeb.jpg)

After a while working with k8s I've coined a term "working in Yaml mines". When dealing with Ops, and in particular k8s, everything seems to be Yaml based. Yaml input, Yaml code, Yaml DSLs, Yaml templates, layering Yaml, combining Yaml. Tedious, error prone, inconvenient.

In RustShop I want to explore a possibility of not touching any Yaml files by hand. After all the core philosophy is: we have two powerful complementary languages: Rust & Nix, we are very familiar with them, and we use them for everything.

I just pushed an initial prototype, so I'm going to briefly present it now.

In RustShop any k8s resources yaml is produced by a "resource generator" binary. Here is [the sample `starter` service resource generator code](https://github.com/rustshop/rustshop/blob/8ea6fafe465fc9f2440db15993236c65b141d689/services/starter-res-gen/src/main.rs#L14) that I've been prototyping on. Resource generator is a normal binary, compiled from a normal Rust code, that can parse command line arguments and environment (if need be) and prints the generated yaml files on `stdout`. To make it easy, it builds on top of a common utility library shared between all resource generator binaries.

I've been familiar for quite some time with a [kube-rs project](https://github.com/kube-rs/kube-rs) that builds  infrastructure for Rust interacting with k8s. It uses a [k8s-openapi](https://github.com/Arnavion/k8s-openapi/) library that is basically a machine-generatated Rust code from k8s swagger type descriptors. Today I've [submitted a PR adding bunch of extra DSL-like functionality to make direct interaction with k8s resource types much more convenient](https://github.com/Arnavion/k8s-openapi/pull/120).

Here is the initial example I've been working on:

```rust
    fn generate(&mut self, ctx: &mut GenContext, _opts: &Self::Opts) -> GenResult<()> {
        let app_name = "starter";
        let pod_selector = ctx.new_labels().insert("app", app_name);
        let labels = pod_selector.clone().insert("shop", "rustshop");

        ctx.add_service(app_name, |s| {
            s.metadata_with(|m| {
                m.labels()
                    .insert("some_extra_service_label".into(), "iamaservice".into());
            })
            .spec()
            .ports_with(|ports| {
                ports.push(ServicePort {
                    port: i32::from(common_app::DEFAULT_LISTEN_PORT),
                    ..Default::default()
                })
            })
            .selector_set(pod_selector.clone());
        })
        .add_deployment(app_name, |d| {
            d.metadata_with(|m| {
                m.labels_insert_from(&labels);
                m.labels()
                    .insert("some_extra_label".into(), "some_extra_value".into());
            })
            .spec()
            .replicas_set(1)
            .selector_set(pod_selector.clone())
            .template_with(|t| {
                t.metadata().labels_insert_from(&labels);
                t.metadata().name_set(app_name.to_owned());
                t.spec().containers_push_with(|c| {
                    c.image_set("redis".to_owned()).name_set("main");
                });
            });
        });

        Ok(())
    }
```

All it takes now to turn it into yaml is running:

```
> nix run .#res.starter
warning: Git tree '/home/dpc/lab/rustshop' is dirty
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app: starter
    some_extra_label: some
  name: starter
spec:
  ports:
    - port: 3000
  selector:
    app: starter

---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    app: starter
    shop: rustshop
    some_extra_label: some_extra_value
  name: starter
spec:
  replicas: 1
  selector:
    matchLabels:
      app: starter
  template:
    metadata:
      labels:
        app: starter
        shop: rustshop
      name: starter
    spec:
      containers:
        - image: redis
          name: main
```

It might not look like this semi-DSL in Rust is much of an improvement at first. 
I have to admit that Rust is not the strongest language for building semi-DSLs like this. There's quite some of `||`s and `{ }`. More than I'd get e.g. in Scala. Yet, I'm very content with these initial results.

It's important to understand that this is most naive an verbatime code, that I came up with just to kick the tires of this thing. Eventually, I'd expect most resource generation libraries to look more like:

```rust
    fn generate(&mut self, ctx: &mut GenContext, _opts: &Self::Opts) -> GenResult<()> {
        let app_name = "zoombo-server";

        let secret_ref = ctx.add_secret(app_name)
        ctx.add_standard_service_with(app_name)
            .add_standard_deploymenet_with(app_name, |d| {
                 d.replicas_set(3);
                 d.add_env_secret_ref(secret_ref);
            });
    }
```

or whatever else we come up with. Abstractions are a double-edged sword, but in my opinion a well-thought abstractions are a key factor that will make this system extremely powerful and effective.

Here are some benefits:

* If you know Rust you don't have to learn weird and often arbitrary ways of how `helm` or `kustomize` chew through yaml files. And after working with these for years now - they all suck, and come with hundreds of papercuts. Not having to use them is a benefit on it's own.
* The whole thing is just plain Rust - a great general purpose language:
  * It's really hard to make silly mistake here.
  * LSP-based autocompletion is always available so you don't have to type a lot of these things.
  * Your IDE will point out mistakes in real time.
  * You can literal one-click/press "jump to definition" and start reading documentation of each field/ resource type both form OpenAPI definition, and any abstractions built on top of them.
  * `cargo doc --open` will show API documentation for all these APIs and utilities.
* Handful of lines might generate hundreds of lines of yaml.
* Viewing the generated output is always just one `nix run ...` away.
* Policies w.r.t certain things can be enforced centrally if so desired.
* Tests can be written for any part of this system.

I am hoping that more people will find this idea at least promising, and I can get this `k8s-openapi` PR merged into mainline. Then I'm planning to keep building RustShop infrastructure around it, and post most interesting bits on this forum, so hit the Subscribe button if you find it interesting!

**Comment by empath-nirvana** [comment]
I think I would like something like this, but I don't think it belongs in k8s-openapi itself, but rather as an opinionated library that imports k8s-openapi.

**Reply by dpc** [reply]
Such a library would have to redo almost everything that `k8s-openapi` does and then will have to do it as extension traits, requiring imports of additional traits for these methods to work or have duplicate types not compatible with ones in `k8s-openapi`.

Having it as an opt-in feature on `k8s-openapi` seems much better for everyone, with the only downsides that maintainers will have to now maintain a bit of extra code (not much, if you see the PR).
